nerfstudio/field_components/encodings.py

```diff
@@ -474,6 +474,91 @@ def upsample_grid(self, resolution: int) -> None:
+    and symmetrical, unlike with VM decomposition where we needed one component with a vector along all the x, y, z
+    directions for symmetry.
+
+    This can be thought of as 3 planes of features perpendicular to the x, y, and z axes, respectively and intersecting
+    at the origin, and the encoding being the element-wise product of the element at the projection of [i, j, k] on
+    these planes.
+
+    The use for this is in representing a tensor decomp of a 4D embedding tensor: (x, y, z, feature_size)
+
+    This will return a tensor of shape (bs:..., num_components)
+
+    Args:
+        resolution: Resolution of grid.
+        num_components: The number of scalar triplanes to use (ie: output feature size)
+        init_scale: The scale of the initial values of the planes
+        product: Whether to use the element-wise product of the planes or the sum
+    """
+
+    plane_coef: TensorType[3, "num_components", "resolution", "resolution"]
+
+    def __init__(
+        self,
+        resolution: int = 32,
+        num_components: int = 64,
+        init_scale: float = 0.1,
+        product: bool = False,
+    ) -> None:
+        super().__init__(in_dim=3)
+
+        self.resolution = resolution
+        self.num_components = num_components
+        self.init_scale = init_scale
+        self.product = product
+
+        self.plane_coef = nn.Parameter(
+            self.init_scale * torch.randn((3, self.num_components, self.resolution, self.resolution))
+        )
+
+    def get_out_dim(self) -> int:
+        return self.num_components
+
+    def forward(self, in_tensor: TensorType["bs":..., 3]) -> TensorType["bs":..., "num_components", "featuresize"]:
+        """Sample features from this encoder. Expects in_tensor to be in range [0, resolution]"""
+
+        original_shape = in_tensor.shape
+        in_tensor = in_tensor.reshape(-1, 3)
+
+        plane_coord = torch.stack([in_tensor[..., [0, 1]], in_tensor[..., [0, 2]], in_tensor[..., [1, 2]]], dim=0)
+
+        # Stop gradients from going to sampler
+        plane_coord = plane_coord.detach().view(3, -1, 1, 2)
+        plane_features = F.grid_sample(
+            self.plane_coef, plane_coord, align_corners=True
+        )  # [3, num_components, flattened_bs, 1]
+
+        if self.product:
+            plane_features = plane_features.prod(0).squeeze(-1).T  # [flattened_bs, num_components]
+        else:
+            plane_features = plane_features.sum(0).squeeze(-1).T
+
+        return plane_features.reshape(*original_shape[:-1], self.num_components)
+
+    @torch.no_grad()
+    def upsample_grid(self, resolution: int) -> None:
+        """Upsamples underlying feature grid
+
+        Args:
+            resolution: Target resolution.
+        """
+        plane_coef = F.interpolate(
+            self.plane_coef.data, size=(resolution, resolution), mode="bilinear", align_corners=True
+        )
+
+        # TODO(ethan): are these torch.nn.Parameters needed?
```

Comment:
Can we resolve this?

Reply:
```
Printing profiling stats, from longest to shortest duration in seconds
Trainer.train_iteration: 0.0282              
VanillaPipeline.get_train_loss_dict: 0.0249              
Traceback (most recent call last):
  File "/home/jake-austin/miniconda3/envs/nerfstudio/bin/ns-train", line 8, in <module>
    sys.exit(entrypoint())
  File "/home/jake-austin/nerfstudio/scripts/train.py", line 247, in entrypoint
    main(
  File "/home/jake-austin/nerfstudio/scripts/train.py", line 233, in main
    launch(
  File "/home/jake-austin/nerfstudio/scripts/train.py", line 172, in launch
    main_func(local_rank=0, world_size=world_size, config=config)
  File "/home/jake-austin/nerfstudio/scripts/train.py", line 87, in train_loop
    trainer.train()
  File "/home/jake-austin/nerfstudio/nerfstudio/engine/trainer.py", line 207, in train
    callback.run_callback_at_location(step, location=TrainingCallbackLocation.AFTER_TRAIN_ITERATION)
  File "/home/jake-austin/nerfstudio/nerfstudio/engine/callbacks.py", line 103, in run_callback_at_location
    self.run_callback(step=step)
  File "/home/jake-austin/nerfstudio/nerfstudio/engine/callbacks.py", line 93, in run_callback
    self.func(*self.args, **self.kwargs, step=step)
  File "/home/jake-austin/nerfstudio/nerfstudio/models/tensorf.py", line 128, in reinitialize_optimizer
    self.field.density_encoding.upsample_grid(resolution)
  File "/home/jake-austin/miniconda3/envs/nerfstudio/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/jake-austin/nerfstudio/nerfstudio/field_components/encodings.py", line 558, in upsample_grid
    self.plane_coef = plane_coef
  File "/home/jake-austin/miniconda3/envs/nerfstudio/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1290, in __setattr__
    raise TypeError("cannot assign '{}' as parameter '{}' "
TypeError: cannot assign 'torch.cuda.FloatTensor' as parameter 'plane_coef' (torch.nn.Parameter or None expected)
```

This is the error thrown if you don't use an nn.Parameter around the plane_coef. I can look into this further if you feel it's worth trying to debug and see if we can get around, but idk if it's super worth trying to debug. I can remove the TODO if you want.

Reply:
Yea, remove the TODO